Add a version of get_or_create_string_literal_symbol that takes a string parameter

antlechner opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->
No functional change, only refactoring.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- n/a My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

antlechner committed:
Move string symbol creation Doxygen to header

Following the new coding standard.

antlechner committed:
Add string symbol creation function that takes a string

This simplifies the code in a few places, since we don't have to repeat
the same lines of constructing the string literal exprt every time.

smowton commented on jbmc/src/java_bytecode/java_string_literals.h:
```diff
@@ -12,9 +12,25 @@ Author: Chris Smowton, chris.smowton@diffblue.com
+///   structure will also be initialised and added to the symbol table.
+/// \return a symbol_expr corresponding to the new or existing literal symbol.
 symbol_exprt get_or_create_string_literal_symbol(
   const exprt &string_expr,
```
Are there any users of this overload left?

antlechner replied:
Only one:
https://github.com/diffblue/cbmc/blob/7dd720b5db62d05bf814b4428c90dcca816d425a/jbmc/src/java_bytecode/java_bytecode_language.cpp#L424

smowton replied:
In that case it would be a very welcome addition to this PR to make `java_string_literal_exprt` rather than directly twiddling operands / named subexpressions -- fair enough though if you think that's out of scope.

antlechner replied:
It looks like there aren't any `java[\w_]+exprt` types yet. Should such a type go into `jbmc/src/java_bytecode/java_types.h` or a new file, e.g. something like `jbmc/src/java_bytecode/java_expressions.h` or similar?
Or does it make more sense to create a new folder `jbmc/src/util`? In that case I would definitely make a new PR for that.

romainbrenguier replied:
Does symex have to handle these expressions in a specific way? If it's the case then they shouldn't be defined in a `jbmc` subdirectory. Otherwise `java_expr.h` would be clearer for showing the class inherits from `exprt` and I'd prefer one class per file, so `java_string_literal_expr.h` would be better.

tautschnig replied:
Hearing about language-specific expressions makes me panic a bit. You will need to make sure all of these are gone by the time the code leaves the Java front-end.

smowton replied:
They are; these are just descriptions of the exprts produced by the bytecode parser. The only other example I'm aware of now, and which I suggest we emulate, is `fieldref_exprt`, which is defined in `java_bytecode_parse_tree.h`.

smowton replied:
(that expression is also local to the parser and becomes `member_exprt` later once we have more information about what it's referring to)

antlechner replied:
I added the new class as you suggested, in a new commit.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 78d181a3c313124a4add32f4690ddd80a34f317b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107812797

antlechner committed:
Create java_string_literal_exprt class

With this new class, we can avoid directly accessing named
subexpressions of exprts with id() == ID_java_string_literal.

smowton reviewed (approved):
Great, thanks a lot for this!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9e499d53dbe32583b5c19f4acb90172e1fbe6e81).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108126079
